docs/r1cs-docs-example.md

```diff
@@ -0,0 +1,530 @@
+# 
+#         // Make multipliers for y from i == [0, k-3]
+#         let first_muly_out = (0..k - 2).rev().fold(last_muly_out, |prev_out, i| {
+#             let (_, _, o) = cs.multiply(prev_out.into(), y[i] - z);
+#             o
+#         });
+# 
+#         // Constrain last x mul output and last y mul output to be equal
+#         cs.constrain(first_mulx_out - first_muly_out);
+#     }
+# }
+impl ShuffleProof {
+    /// Attempt to construct a proof that `output` is a permutation of `input`.
+    ///
+    /// Returns a tuple `(proof, input_commitments || output_commitments)`.
+    pub fn prove<'a, 'b>(
+        pc_gens: &'b PedersenGens,
+        bp_gens: &'b BulletproofGens,
+        transcript: &'a mut Transcript,
+        input: &[Scalar],
+        output: &[Scalar],
+    ) -> Result<(ShuffleProof, Vec<CompressedRistretto>), R1CSError> {
+        // Apply a domain separator with the shuffle parameters to the transcript
+        let k = input.len();
+        transcript.commit_bytes(b"dom-sep", b"ShuffleProof");
+        transcript.commit_bytes(b"k", Scalar::from(k as u64).as_bytes());
+
+        // Collect witness assignments
+        let mut v = Vec::with_capacity(2 * k);
+        v.extend_from_slice(input);
+        v.extend_from_slice(output);
+
+        // Construct blinding factors using a TranscriptRng
+        // Note: a non-example implementation would want to operate on existing commitments
+        let mut rng = {
+            let mut builder = transcript.build_rng();
+            // commit the secret values
+            for &v_i in &v {
+                builder = builder.commit_witness_bytes(b"v_i", v_i.as_bytes());
+            }
+            use rand::thread_rng;
+            builder.finalize(&mut thread_rng())
+        };
+        let v_blinding: Vec<Scalar> = (0..2 * k).map(|_| Scalar::random(&mut rng)).collect();
+
+        // Construct a `ConstraintSystem` instance for the shuffle gadget
+        let (mut cs, variables, commitments) =
+            ProverCS::new(&bp_gens, &pc_gens, transcript, v, v_blinding.clone());
+
+        // Allocate variables and add constraints to the constraint system
+        let (input_vars, output_vars) = variables.split_at(k);
+        ShuffleProof::gadget(&mut cs, input_vars, output_vars);
+
+        // Generate proof
+        let proof = cs.prove()?;
+
+        Ok((ShuffleProof(proof), commitments))
+    }
+}
+```
+
+## Verifiying a proof
+
+The verifier receives a proof, and a list of committed inputs and outputs, from the prover. It passes these to the `verify` function, and receives a `Result::ok()` if the proof verified correctly and a `Result::error(R1CSError)` otherwise. The proof could fail to verify for a variety of reasons - if the original scalar outputs were not a valid reordering of the inputs, if the commitments were generated or transmitted incorrectly, or if the constraint system was built differently for the prover and verifier. 
```

Comment:
It might be simpler just to say that the proof will verify, or not, since the user shouldn't have to know all the reasons why the proof could fail (and we don't expose them in the error anyways, on purpose)